monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/UnfoldEvalObservable.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2014-2019 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.reactive.internal.builders
+
+import monix.eval.Task
+import monix.execution.Ack.{Continue, Stop}
+import monix.execution.{Callback, Cancelable}
+import monix.reactive.Observable
+import monix.reactive.observers.Subscriber
+
+import scala.util.control.NonFatal
+
+private[reactive] final class UnfoldEvalObservable[S, A](seed: S, f: S => Task[Option[(A, S)]]) extends Observable[A] {
+
+  def unsafeSubscribeFn(subscriber: Subscriber[A]): Cancelable = {
+    import subscriber.scheduler
+    var streamErrors = true
+    try {
+      val init = seed
+      streamErrors = false
+
+      Task
+        .defer(loop(subscriber, init))
+        .executeWithOptions(_.enableAutoCancelableRunLoops)
+        .runAsync(Callback.empty)
+    } catch {
+      case ex if NonFatal(ex) =>
+        if (streamErrors) subscriber.onError(ex)
+        else subscriber.scheduler.reportFailure(ex)
+        Cancelable.empty
+    }
+  }
+
+  def loop(subscriber: Subscriber[A], state: S): Task[Unit] =
+    try f(state).redeemWith(
+      { ex =>
+        subscriber.onError(ex)
+        Task.unit
+      }, {
+        case Some((a, newState)) =>
+          Task.fromFuture(subscriber.onNext(a)).flatMap {
+            case Continue =>
+              loop(subscriber, newState)
+            case Stop =>
+              subscriber.onComplete()
```

Comment:
We shouldn't call `onComplete()` here because downstream subscriber already returned `Stop` to us and most likely called `onComplete` on its own

See: https://monix.io/docs/3x/reactive/observers.html#contract